superset-frontend/src/explore/components/controls/DndColumnSelectControl/DndFilterSelect.tsx

```diff
@@ -384,14 +405,21 @@ const DndFilterSelect = (props: DndFilterSelectProps) => {
 
   const canDrop = useCallback(
     (item: DatasourcePanelDndItem) => {
+      if (
+        extra.disallow_adhoc_metrics &&
+        (item.type !== DndItemType.Column ||
+          !availableColumnSet.has((item.value as ColumnMeta).column_name))
+      ) {
+        return false;
+      }
+
       if (item.type === DndItemType.Column) {
-        return props.columns.some(
-          col => col.column_name === (item.value as ColumnMeta).column_name,
-        );
+        const columnName = (item.value as ColumnMeta).column_name;
+        return availableColumnSet.has(columnName);
```

Comment:
@kgabryje I made a change from #27209 to improve the array iteration check by utilizing a hashset, which offers better performance and optimization.

Reply:
Makes sense, thank you!